lib/game.rb

```diff
@@ -0,0 +1,49 @@
+class Game
+    include Interface
+
+    def initialize(player1, player2, board)
+        @player1 = player1
+        @player2 = player2
+        @board = board
+        @status = "initial"
+    end
+
+    def play
+        getInput('welcome')      
+        loop do   
+            restart if @status == "initial"
+            until game_finish? 
+                next_move(@player1) if !game_finish?
+                break if game_finish?
+                next_move(@player2) if !game_finish?
+            end 
+            break if !play_again?
+        end
+        finish 
+    end
+
+    private
+
+    def restart
+        @player1.restartPlayer
+        @player2.restartPlayer
+        @board.restartBoard
+        @status = "continue"               
+        show(@board.positions)  
+    end
+  
+    def game_finish?
+       has_won?(@player1)|| has_won?(@player2) || @board.full?
+    end
+
+    def has_won?(player)
```

Comment:
Three things about this `has_won?` method:

1. It still has some UI code in it (`print`).
2. As implied by the `?` in the method name, the method should simply be returning `true` or `false` and not do anything else. We can extract the parts of the method that don't do this and place those in the UI module.
3. You guys have already nicely designed the `board.win?` method, probably we don't need the `has_won?` anymore. We can therefore just write our `game_finish?` method as:

```ruby
def game_finish?
  @board.win?(@player1.inputs) || @board.win?(@player2.inputs) || @board.full?
end
```
We can then transfer the following into the UI module:

```ruby
# place this in the UI module
def winner_display(player, board)
  puts "#{player.name}"
  getInput("win")
  show(board.positions)
end
```

Ok this is as much clue as I can give you guys. How you invoke this `winner_display` method here in the `Game` class I will leave to you :)

Reply:
ok, we figure it out now, thanks!